akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,643 @@
+   * @param perProducerBufferSize Buffer space used per producer.
+   */
+  def source[T](perProducerBufferSize: Int): Source[T, Sink[T, NotUsed]] =
+    Source.fromGraph(new MergeHub[T](perProducerBufferSize))
+
+  final class ProducerFailed(msg: String, cause: Throwable) extends RuntimeException(msg, cause)
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class MergeHub[T](perProducerBufferSize: Int) extends GraphStageWithMaterializedValue[SourceShape[T], Sink[T, NotUsed]] {
+  require(perProducerBufferSize > 0, "Buffer size must be positive")
+
+  val out: Outlet[T] = Outlet("MergeHub.out")
+  override val shape: SourceShape[T] = SourceShape(out)
+
+  // Half of buffer size, rounded up
+  private[this] val DemandThreshold = (perProducerBufferSize / 2) + (perProducerBufferSize % 2)
+
+  private sealed trait Event {
+    def id: Long
+  }
+
+  private final case class Element(id: Long, elem: T) extends Event
+  private final case class Register(id: Long, demandCallback: AsyncCallback[Long]) extends Event
+  private final case class Deregister(id: Long) extends Event
+
+  final class InputState(signalDemand: AsyncCallback[Long]) {
+    private var untilNextDemandSignal = DemandThreshold
+
+    def onElement(): Unit = {
+      untilNextDemandSignal -= 1
+      if (untilNextDemandSignal == 0) {
+        untilNextDemandSignal = DemandThreshold
+        signalDemand.invoke(DemandThreshold)
+      }
+    }
+
+    def close(): Unit = signalDemand.invoke(MergeHub.Cancel)
+
+  }
+
+  final class MergedSourceLogic(_shape: Shape, producerCount: AtomicLong) extends GraphStageLogic(_shape) with OutHandler {
+    /*
+     * Basically all merged messages are shared in this queue. Individual buffer sizes are enforced by tracking
+     * demand per producer in the 'demands' Map. One twist here is that the same queue contains control messages,
+     * too. Since the queue is read only if the output port has been pulled, downstream backpressure can delay
+     * processing of control messages. This causes no issues though, see the explanation in 'tryProcessNext'.
+     */
+    private val queue = new AbstractNodeQueue[Event] {}
+    @volatile private[this] var needWakeup = false
+    @volatile private[this] var shuttingDown = false
+
+    private[this] val demands = scala.collection.mutable.LongMap.empty[InputState]
+    private[this] val wakeupCallback = getAsyncCallback[NotUsed]((_) ⇒
+      // We are only allowed to dequeue if we are not backpressured. See comment in tryProcessNext() for details.
+      if (isAvailable(out)) tryProcessNext(firstAttempt = true)
+    )
+
+    setHandler(out, this)
+
+    // Returns true when we have not consumed demand, false otherwise
+    private def onEvent(ev: Event): Boolean = ev match {
+      case Element(id, elem) ⇒
+        demands(id).onElement()
+        push(out, elem)
+        false
+      case Register(id, callback) ⇒
+        // TODO: Check for already existing state? Should not happen if works correctly though
```

Comment:
How would it happen that we get a Register with an already used id here? Just fail the stage with IllegalState in that case? Or just remove the TODO because it cannot happen.